commons/src/main/java/io/aiven/kafka/tieredstorage/commons/storage/s3/S3StorageConfig.java

```diff
@@ -0,0 +1,204 @@
+                ConfigDef.Importance.LOW,
+                S3_CREDENTIALS_PROVIDER_CLASS_DOC)
+            .define(
+                S3_ACCESS_KEY_ID_CONFIG,
+                ConfigDef.Type.PASSWORD,
+                null,
+                new NonEmptyPassword(),
+                ConfigDef.Importance.MEDIUM,
+                S3_ACCESS_KEY_ID_DOC)
+            .define(
+                S3_SECRET_ACCESS_KEY_CONFIG,
+                ConfigDef.Type.PASSWORD,
+                null,
+                new NonEmptyPassword(),
+                ConfigDef.Importance.MEDIUM,
+                S3_SECRET_ACCESS_KEY_DOC);
+    }
+
+    S3StorageConfig(final Map<String, ?> props) {
+        super(CONFIG, props);
+    }
+
+    AmazonS3 s3Client(final AWSCredentialsProvider credentialsProvider) {
+        AmazonS3ClientBuilder s3ClientBuilder = AmazonS3ClientBuilder.standard();
+        final String s3ServiceEndpoint = getString(S3_ENDPOINT_URL_CONFIG);
+        final String region = getString(S3_REGION_CONFIG);
+        if (s3ServiceEndpoint == null) {
+            s3ClientBuilder = s3ClientBuilder.withRegion(region);
+        } else {
+            final AwsClientBuilder.EndpointConfiguration endpointConfiguration =
+                new AwsClientBuilder.EndpointConfiguration(s3ServiceEndpoint, region);
+            s3ClientBuilder = s3ClientBuilder.withEndpointConfiguration(endpointConfiguration)
+                .withPathStyleAccessEnabled(true);
+        }
+        s3ClientBuilder.setCredentials(credentialsProvider);
+        return s3ClientBuilder.build();
+    }
+
+    public AWSCredentialsProvider credentialsProvider() {
+        try {
+            @SuppressWarnings("unchecked") final Class<? extends AWSCredentialsProvider> providerClass =
+                (Class<? extends AWSCredentialsProvider>) getClass(S3_CREDENTIALS_PROVIDER_CLASS_CONFIG);
+            final boolean areCredentialsProvided =
+                getPassword(S3_ACCESS_KEY_ID_CONFIG) != null && getPassword(S3_SECRET_ACCESS_KEY_CONFIG) != null;
+            if (areCredentialsProvided || providerClass.isAssignableFrom(AWSStaticCredentialsProvider.class)) {
+                if (!areCredentialsProvided) {
+                    throw new ConfigException("When "
+                        + AWSStaticCredentialsProvider.class.getName()
+                        + " is used, "
+                        + S3_ACCESS_KEY_ID_CONFIG
+                        + " and "
+                        + S3_SECRET_ACCESS_KEY_CONFIG
+                        + " must be provided.");
+                }
+                final AWSCredentials staticCredentials = new BasicAWSCredentials(
+                    getPassword(S3_ACCESS_KEY_ID_CONFIG).value(),
+                    getPassword(S3_SECRET_ACCESS_KEY_CONFIG).value()
+                );
+                return new AWSStaticCredentialsProvider(staticCredentials);
+            }
+            return providerClass.getDeclaredConstructor().newInstance();
```

Comment:
Consider using `getConfiguredInstance`. But no biggie, it just does some checks and exception handling for you.